### DISCUSSION TITLE
Access to Chat history in semantic kernel function

### DISCUSSION THREAD

**Discussion by JP-droidic** [microsoft/semantic-kernel#7186] [Q&A] [answered]
Hi all

is it possible to access/retrieve chat history in autocalled SK function?

**Comment by moonbox3** [comment, marked as answer]
Hi @JP-droidic, yes, the chat history is part of the accessible metadata. Can you please let us know which SK language you're using and I can provide more detail. Thanks.

**Reply by JP-droidic** [reply]
@moonbox3  C# please

**Reply by moonbox3** [reply]
Hi @JP-droidic, apologies for the delay in responding. This is currently available in Python today, but not yet in C#. The following task is tracking the work: https://github.com/microsoft/semantic-kernel/issues/7263

**Reply by MaxiPigna** [reply]
> @moonbox3 C# please

Did you try to add it to the Kernel Data and access the history from there?

**Reply by dmytrostruk** [reply]
@MaxiPigna Thanks for providing this suggestion! I think it should work as a temporary solution, but we are investigating if it's possible to inject ChatHistory to the function directly, so there will be no need in using `kernel.Data` and casting object to ChatHistory. We will provide an update here as soon as we make a progress on this. Thank you!

**Reply by huislaw** [reply]
> Hi @JP-droidic, apologies for the delay in responding. This is currently available in Python today, but not yet in C#. The following task is tracking the work: #7263

How do I access it using Python ?
Thanks.

**Comment by dmytrostruk** [comment]
@JP-droidic Just to double check, you want to have access to `ChatHistory` object inside kernel function, when it's invoked, is that correct? It would be great if you can share some code snippet of desired functionality, if possible. Thanks!

**Reply by JP-droidic** [reply]
That is correct @dmytrostruk. This is our UC:

We have a semantic function that searches Azure AI search and returns a set of relevant articles back to the LLM - basically a classic RAG scenario. This is done as an SK function as this is just one of the things a user might ask, and potentially, a user can say:

Give 10 latest articles about XYZ and put them into a PDF document:

In this case, we would invoke our Azure AI search, followed by an SK function that generates PDFs.

The issue here is that the SK function for Azure AI is quite complex; we build a filter with 10ish parameters (some of the parameters can be data range, content type, authors, keyword, order, number of results ...)

So what we have on the SK function is:

1. Parsing user's request to extract all search parameters -> for this we use a call to LLM
2. Creating Azure AI search filter
3. Executing Azure AI search
4. Postprocessing Azure AI search results
5. return a list of results back

Our main semantic function accepts user requests as a parameter, which we pass to the SK function's LLM to extract search parameters (no 1).

When we have a conversation with multiple back and forth e.g.:

Call 1:
Give 10 latest articles about XYZ and put them into a PDF document:
Call 2 can you focus on just articles published in last 2 months
Call 3 filter also by author XXX

The input we get during call 3 usually doesn't contain everything the user is looking for. So this is the moment where we need full history access inside SK function so we can pass on full history to the LLM to be able to extra all the parameters we use for Azure Ai Search:
no of result: 10 
content-type: articles 
keyword: XYZ
date range today- 2 months
Authror= XXX

hope this explains

thanks

**Reply by dmytrostruk** [reply]
@JP-droidic Got it, thanks for additional information! Have you considered to maintain `ChatHistory` on your side and invoke your requests in a loop? Like in this example:
https://github.com/microsoft/semantic-kernel/blob/305b2ee28b6d119b9e2fa63bd1201a90e2b9a29b/dotnet/samples/LearnResources/MicrosoftLearn/CreatingFunctions.cs#L48-L72

This is just a demonstration, if you have a web server, you can serialize and store chat history, and then re-use it for further requests, but the main point here, based on your example, you should be able to achieve it if you store chat history on your side and then re-use it with additional requests, so the model will better understand the context. It's still possible to enable auto function calling and invoke Azure AI Search related operations, that are wrapped in SK functions, if needed. Please let us know if it's something that can handle your scenario, thanks!

**Reply by JP-droidic** [reply]
yes, that makes sense. Thank you for your help
